### DISCUSSION TITLE
Mend Renovate failing to authorise with Docker Hub

### DISCUSSION THREAD

**Discussion by CMarcher-Ahuora** [renovatebot/renovate#29320] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitHub, 37.381.7

### Please tell us more about your question or problem

I am having a great deal of trouble attempting to define a valid Docker Hub `hostRule` entry that allows Renovate to authenticate with a personal Docker Hub account to read from a few **private** repositories.

This is the general error returned by Mend Renovate: `Renovate failed to look up the following dependencies`, and for each failed image dependency: `Could not determine new digest for update`.

Here is one variant of the config I have tried:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended"
  ],
  "hostRules": [
    {
      "hostType": "docker",
      "username": "<redacted>",
      "encrypted": {
        "password" : "<redacted>" 
      }
    }
  ]
}
```

I have tried to:
- replace the `hostType` field with `"matchHost" : "docker.io"`
- include both `hostType` and `matchHost`
- include neither
- include them as separate rules with the same username and password

I'm using [the Renovate encrypt app](https://app.renovatebot.com/encrypt) to encrypt the Docker Hub access token.

Attached is one of the HTTP responses from the debug log on the Mend Renovate portal.

If I use `hostType`, that specific 401 Unauthorised response is seen in the log, while no HTTP response is written to the log when using `matchHost`.

What am I doing wrong?

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: Unauthorized docker lookup
{
  "registryHost": "https://index.docker.io"
  "dockerRepository": "user/repo"
}

DEBUG: Response code 401 (Unauthorized)
{
  "err": {
    "name": "HTTPError",
    "code": "ERR_NON_2XX_3XX_RESPONSE",
    "timings": {
      "start": 1716976767751,
      "socket": 1716976767752,
      "lookup": 1716976767757,
      "connect": 1716976767766,
      "secureConnect": 1716976767769,
      "upload": 1716976767771,
      "response": 1716976767776,
      "end": 1716976767777,
      "phases": {
        "wait": 1,
        "dns": 5,
        "tcp": 9,
        "tls": 3,
        "request": 2,
        "firstByte": 5,
        "download": 1,
        "total": 26
      }
    },
    "message": "Response code 401 (Unauthorized)",
    "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/local/renovate/node_modules/.pnpm/got@11.8.6/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
    "options": {
      "headers": {
        "user-agent": "RenovateBot/37.377.8 (https://github.com/renovatebot/renovate)",
        "authorization": "***********",
        "accept": "application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.docker.distribution.manifest.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://index.docker.io/v2/user/repo/manifests/latest",
      "hostType": "docker",
      "username": "",
      "password": "",
      "method": "HEAD",
      "http2": false
    },
    "response": {
      "statusCode": 401,
      "statusMessage": "Unauthorized",
      "body": "",
      "headers": {
        "content-type": "application/json",
        "docker-distribution-api-version": "registry/2.0",
        "www-authenticate": "Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\",scope=\"repository:user/repo:pull\",error=\"insufficient_scope\"",
        "date": "Wed, 29 May 2024 09:59:27 GMT",
        "content-length": "155",
        "strict-transport-security": "max-age=31536000",
        "docker-ratelimit-source": "b7bd6acf-c9a5-4a04-b9cc-05f2c11bef8a",
        "connection": "close"
      },
      "httpVersion": "1.1",
      "retryCount": 0
    }
  }
}

```

</details>

**Comment by rarkins** [comment]
@CMarcher-Ahuora FYI the hosted app tries using the hub.docker.com API first, but I forget if it works with private packages or needs to fall back to index.docker.io. Do you see evidence that both are being tried?

Additionally the hosted app includes a read-only Docker Hub token in jobs to avoid rate limiting, so it's possible that our built-in token is taking precedence over your token. Try setting matchHost=auth.docker.io and matchHost=index.docker.io (i.e. two hostRules) so that it's a longer match than our matchHost=docker.io. I would have expected your rule to take precedence over our global rules though.

**Comment by CMarcher-Ahuora** [comment]
@rarkins Doesn't seem to want to work with your suggested hostRules either. I also had an attempt at using a temporary raw access token as the password (instead of using the Renovate encrypt tool) with either matchHost or hostType rule variants, but that didn't result in a better outcome.

As according to another snippet of the logs, it seems that there is a reference to both index.docker.io and auth.docker.io.

```
DEBUG: getDigest(https://index.docker.io, company/repo, undefined)
DEBUG: getDigest(https://index.docker.io, company/repo, undefined)
DEBUG: getDigest(https://index.docker.io, company/repo, undefined)
DEBUG: getManifestResponse(https://index.docker.io, company/repoi, latest, head)
DEBUG: getManifestResponse(https://index.docker.io, company/repo, latest, head)
DEBUG: getManifestResponse(https://index.docker.io, company/repo, latest, head)
DEBUG: hostRules: basic auth for https://index.docker.io
DEBUG: HEAD https://index.docker.io/v2/company/repo/manifests/latest = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=43)
DEBUG: Unauthorized docker lookup
{
  "registryHost": "https://index.docker.io"
  "dockerRepository": "company/repo"
}

DEBUG: Response code 401 (Unauthorized)
{
  "err": {
    "name": "HTTPError",
    "code": "ERR_NON_2XX_3XX_RESPONSE",
    "timings": {
      "start": 1717133540337,
      "socket": 1717133540337,
      "lookup": 1717133540337,
      "connect": 1717133540345,
      "secureConnect": 1717133540351,
      "upload": 1717133540356,
      "response": 1717133540380,
      "end": 1717133540380,
      "phases": {
        "wait": 0,
        "dns": 0,
        "tcp": 8,
        "tls": 6,
        "request": 5,
        "firstByte": 24,
        "download": 0,
        "total": 43
      }
    },
    "message": "Response code 401 (Unauthorized)",
    "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/local/renovate/node_modules/.pnpm/got@11.8.6/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
    "options": {
      "headers": {
        "user-agent": "RenovateBot/37.377.8 (https://github.com/renovatebot/renovate)",
        "authorization": "***********",
        "accept": "application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.docker.distribution.manifest.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://index.docker.io/v2/company/repo/manifests/latest",
      "hostType": "docker",
      "username": "",
      "password": "",
      "method": "HEAD",
      "http2": false
    },
    "response": {
      "statusCode": 401,
      "statusMessage": "Unauthorized",
      "body": "",
      "headers": {
        "content-type": "application/json",
        "docker-distribution-api-version": "registry/2.0",
        "www-authenticate": "Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\",scope=\"repository:company/repo:pull\",error=\"insufficient_scope\"",
        "date": "Fri, 31 May 2024 05:32:20 GMT",
        "content-length": "159",
        "strict-transport-security": "max-age=31536000",
        "docker-ratelimit-source": "bc0a62b6-98cb-4fa4-951f-71a373a54d11",
        "connection": "close"
      },
      "httpVersion": "1.1",
      "retryCount": 0
    }
  }
}

```

**Reply by rarkins** [reply]
I think we'll next try to get an example working on our end. Reproduction should be simple:

- Needs a Docker account with a private image, image should have for example two tagged versions 1.0.0 and 1.0.1
- Dockerfile in repo with simple "FROM image:1.0.0" which we expect to be upgraded to 1.0.1
- Provision a hostRule with username/password set
- Verify that the lookup and update is successful

**Reply by rarkins** [reply]
I've been able to verify that Renovate and the app are working exactly as I'd expect.

Successful working reproduction: https://github.com/rarkins/dockerhubprivatetest1

I did this:
- Made my Docker account paid so that I can create private images
- Created a private repository image
- Pushed tags 1.0.0 and 1.0.1
- Created a github repo with Dockerfile referencing this image
- Verified that Renovate failed to look it up initially
- Created a read-only PAT token from Docker Hub portal
- Added Docker PAT as a secret through developer.mend.io
- Created a host rule using that secret: https://github.com/rarkins/dockerhubprivatetest1/blob/6093907e4c418bb28e6cc443301d4189b32cd3ca/renovate.json#L6-L11

And immediately after that, Renovate looked up the image fine. matchHost=docker.io was enough

Could you try to replicate my steps with your own example?

**Reply by CMarcher-Ahuora** [reply, marked as answer]
I replicated your steps with a similar temporary set-up, and it seems to work fine. On the actual repository I was having problems with, it seems defining a secret (as you did in your example) within the Mend Renovate repo config did the trick, and I'm no longer encountering the 401 Unauthorised error.

```
"hostRules": [
    {
      "hostType" : "docker",
      "matchHost": "docker.io",
      "username": "<username>",
      "password" : "{{ secrets.DOCKER_HUB_TOKEN }}"
    }
  ]
```

I wonder, is [this](https://app.renovatebot.com/encrypt) still considered a valid way to use secrets within a Renovate config? I simply could not get it to work as needed for my use-case.

Thanks for taking your time to help with this.

**Reply by rarkins** [reply]
Yes, the encrypted form is still valid but we'll be deprecating it. It's working for thousands of repos
